barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp

```diff
@@ -2850,10 +2954,46 @@ AvmError AvmTraceBuilder::op_emit_nullifier(uint8_t indirect, uint32_t nullifier
 {
     auto const clk = static_cast<uint32_t>(main_trace.size()) + 1;
 
-    auto [row, error] = create_kernel_output_opcode(indirect, clk, nullifier_offset);
-    kernel_trace_builder.op_emit_nullifier(clk, side_effect_counter, row.main_ia);
+    RowWithError row_with_error = create_kernel_output_opcode(indirect, clk, nullifier_offset);
+    Row row = row_with_error.row;
+    AvmError error = row_with_error.error;
     row.main_sel_op_emit_nullifier = FF(1);
 
+    // Do merkle check
+    FF nullifier_value = row.main_ia;
+    FF siloed_nullifier = AvmMerkleTreeTraceBuilder::unconstrained_silo_nullifier(
+        current_public_call_request.contract_address, nullifier_value);
+
+    // This is a little bit fragile - but we use the fact that if we traced a nullifier that already exists (which is
+    // invalid), we would have stored it under a read hint.
+    NullifierReadTreeHint nullifier_read_hint = execution_hints.nullifier_read_hints.at(nullifier_read_counter);
+    bool is_update = merkle_tree_trace_builder.perform_nullifier_read(clk,
+                                                                      nullifier_read_hint.low_leaf_preimage,
+                                                                      nullifier_read_hint.low_leaf_index,
+                                                                      nullifier_read_hint.low_leaf_sibling_path,
+                                                                      intermediate_tree_snapshots.nullifier_tree.root);
+    if (is_update) {
+        // If we are in this branch, then the nullifier already exists in the tree
+        // WE NEED TO RAISE AN ERROR FLAG HERE - for now we do nothing, except increment the counter
+        nullifier_read_counter++;
+    } else {
```

Comment:
Would it be hard to raise error flag here in this PR or a followup?

Reply:
Yeah - that was more a reference to error in the circuit sense tbh, this isnt an exceptional halt for us right? It is a revertible error

Reply:
Maybe there was confusion about terminology. We just want alignment between simulator and witgen. In the simulator for this error, the journal will throw an exception, and AvmSimulator will catch it and return success=false to the caller. I think the equivalent i witgen is to raise an error flag, which i think we accomplish now by returning an error from the opcode function?

Reply:
Yep i've added a new error type to handle duplicate nullifiers